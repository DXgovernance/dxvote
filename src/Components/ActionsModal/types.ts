import { DecodedAction } from 'old-components/Guilds/ActionsBuilder/types';

export interface ActionModalProps {
  isOpen: boolean;
  setIsOpen: (isOpen: boolean) => void;
  onAddAction: (action: DecodedAction) => void;
}
