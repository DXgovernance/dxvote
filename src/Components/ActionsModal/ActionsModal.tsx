import React from 'react';
import { useWeb3React } from '@web3-react/core';
import { BigNumber, utils } from 'ethers';
import { useTranslation } from 'react-i18next';

import { useTypedParams } from 'Modules/Guilds/Hooks/useTypedParams';
import { RichContractData } from 'hooks/Guilds/contracts/useRichContractRegistry';
import {
  defaultValues,
  getEditor,
  supportedActions,
} from 'old-components/Guilds/ActionsBuilder/SupportedActions';
import {
  DecodedAction,
  DecodedCall,
  SupportedAction,
} from 'old-components/Guilds/ActionsBuilder/types';
import { Modal } from 'old-components/Guilds/common/Modal';
import {
  ContractActionsList,
  ApproveSpendTokens,
  ContractsList,
  ParamsForm,
} from './components';
import { EditorWrapper, BlockButton } from './ActionsModal.styled';
import { ActionModalProps } from './types';

const ActionModal: React.FC<ActionModalProps> = ({
  isOpen,
  setIsOpen,
  onAddAction,
}) => {
  const { t } = useTranslation();
  const { guildId } = useTypedParams();
  const { account: walletAddress } = useWeb3React();
  // Supported Actions
  const [selectedAction, setSelectedAction] =
    React.useState<SupportedAction>(null);
  const [selectedActionContract, setSelectedActionContract] =
    React.useState<utils.Interface>(null);

  // Generic calls
  const [selectedContract, setSelectedContract] =
    React.useState<RichContractData>(null);
  const [selectedFunction, setSelectedFunction] = React.useState<string>(null);

  const [data, setData] = React.useState<DecodedCall>(null);
  const [payableFnData, updatePayableFnData] = React.useState<any>(null);

  function getHeader() {
    if (selectedFunction) {
      return selectedContract.functions.find(
        fn => fn.functionName === selectedFunction
      )?.title;
    }

    if (selectedContract) {
      return selectedContract?.title;
    }

    if (selectedAction) {
      return supportedActions[selectedAction].title;
    }

    return t('addAction');
  }

  function getContent() {
    if (selectedFunction) {
      const contractInterface = selectedContract.contractInterface;
      const contractId = selectedContract.contractAddress;
      const fn = selectedContract.functions.find(
        fn => fn.functionName === selectedFunction
      );
      const isPayable: boolean = fn?.spendsTokens;
      // Return approval form if function is marked with spendsTokens=true
      if (isPayable && !payableFnData) {
        return <ApproveSpendTokens onConfirm={updatePayableFnData} />;
      }

      return (
        <ParamsForm
          fn={fn}
          onSubmit={args => {
            onAddAction({
              id: `action-${Math.random()}`,
              contract: contractInterface,
              decodedCall: {
                callType: SupportedAction.GENERIC_CALL,
                from: guildId,
                to: contractId,
                function: contractInterface.getFunction(selectedFunction),
                value: BigNumber.from(0),
                args,
                richData: selectedContract,
              },
              approval: payableFnData,
            });
            handleClose();
          }}
        />
      );
    }

    if (selectedContract) {
      return (
        <ContractActionsList
          contract={selectedContract}
          onSelect={setSelectedFunction}
        />
      );
    }

    if (selectedAction) {
      const Editor = getEditor(selectedAction);
      return (
        <EditorWrapper data-testid="actions-modal-editor">
          <Editor decodedCall={data} updateCall={setData} />
          <BlockButton onClick={saveSupportedAction}>
            {t('saveAction')}
          </BlockButton>
        </EditorWrapper>
      );
    }

    return (
      <ContractsList
        onSelect={setSelectedContract}
        onSupportedActionSelect={setSupportedAction}
      />
    );
  }

  function goBack() {
    if (selectedFunction) {
      setSelectedFunction(null);
      updatePayableFnData(null);
    } else if (selectedContract) {
      setSelectedContract(null);
    } else if (selectedAction) {
      setSelectedAction(null);
      setSelectedActionContract(null);
    }

    setData(null);
  }

  function setSupportedAction(action: SupportedAction) {
    const defaultDecodedAction = defaultValues[action] as DecodedAction;
    if (!defaultDecodedAction) return null;

    defaultDecodedAction.decodedCall.from = guildId;
    defaultDecodedAction.decodedCall.callType = action;
    switch (action) {
      case SupportedAction.REP_MINT:
        defaultDecodedAction.decodedCall.args.to = walletAddress;
        break;
    }
    setData(defaultDecodedAction.decodedCall);
    setSelectedAction(action);
    setSelectedActionContract(defaultDecodedAction.contract);
  }

  function saveSupportedAction() {
    if (!selectedAction || !data || !setSelectedActionContract) return;

    const decodedAction: DecodedAction = {
      id: `action-${Math.random()}`,
      decodedCall: data,
      contract: selectedActionContract,
    };

    onAddAction(decodedAction);
    handleClose();
  }

  const handleClose = () => {
    setSelectedFunction(null);
    setSelectedContract(null);
    setSelectedAction(null);
    setSelectedActionContract(null);
    updatePayableFnData(null);
    setIsOpen(false);
  };

  return (
    <Modal
      isOpen={isOpen}
      onDismiss={handleClose}
      header={getHeader()}
      maxWidth={300}
      backnCross={!!selectedAction || !!selectedContract}
      prevContent={goBack}
    >
      {getContent()}
    </Modal>
  );
};

export default ActionModal;
