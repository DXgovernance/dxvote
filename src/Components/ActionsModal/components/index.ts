export { ContractActionsList } from './ContractActionsList';
export { ApproveSpendTokens } from './ApproveSpendTokens';
export { ContractsList } from './ContractsList';
export { ParamsForm } from './ParamsForm';
