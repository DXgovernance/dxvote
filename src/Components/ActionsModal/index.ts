export { default as ActionModal } from './ActionsModal';
