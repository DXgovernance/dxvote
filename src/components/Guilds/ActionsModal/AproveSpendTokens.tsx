// import { RegistryContract } from 'hooks/Guilds/contracts/useContractRegistry';
import { useState, useMemo } from 'react';
import { SectionTitle, SectionWrapper, Wrapper } from './styles';
import styled from 'styled-components';

import { BlockButton } from '.';
import { FiChevronDown } from 'react-icons/fi';
import { useERC20Info } from 'hooks/Guilds/ether-swr/erc20/useERC20Info';
import Input from 'components/Guilds/common/Form/Input';
import { useTokenList } from 'hooks/Guilds/tokens/useTokenList';
import { useWeb3React } from '@web3-react/core';
import Avatar from 'components/Guilds/Avatar';
import { resolveUri } from 'utils/url';

import { BigNumber } from 'ethers';
import TokenPicker from 'components/Guilds/TokenPicker';
import { Box } from 'components/Guilds/common/Layout';
import TokenAmountInput from 'components/Guilds/common/Form/TokenAmountInput';
const ControlRow = styled(Box)`
  display: flex;
  align-items: stretch;
  height: 100%;
`;

const Control = styled(Box)`
  display: flex;
  flex-direction: column;
  margin: 0.75rem 0;
  width: 100%;
`;

const ControlLabel = styled(Box)`
  margin-bottom: 0.75rem;
`;

const Spacer = styled(Box)`
  margin-right: 1rem;
`;

interface TokensToSpendProps {
  onConfirm: ({ amount, token }: { amount: BigNumber; token: string }) => void;
}

const TokensToSpend: React.FC<TokensToSpendProps> = ({ onConfirm }) => {
  const { chainId } = useWeb3React();

  const [isTokenPickerOpen, setIsTokenPickerOpen] = useState(false);
  const [amount, setAmount] = useState<BigNumber>(null);
  const [token, setToken] = useState<string>(null);

  const { tokens } = useTokenList(chainId);
  const selectedToken = useMemo(() => {
    if (!token || !tokens) return null;

    return tokens.find(({ address }) => address === token);
  }, [tokens, token]);

  // TODO: tokenInfo not working for devscript
  const { data: tokenInfo } = useERC20Info(token);

  const confirm = () => {
    if (!amount || !token) return; // TODO: validate
    onConfirm({
      amount,
      token,
    });
  };

  return (
    <Wrapper>
      <SectionWrapper>
        <SectionTitle>Select tokens to spend</SectionTitle>
        <ControlRow>
          <Control>
            <ControlLabel>Amount</ControlLabel>
            <ControlRow>
              <TokenAmountInput
                decimals={tokenInfo?.decimals}
                value={amount}
                onChange={setAmount}
              />
            </ControlRow>
          </Control>

          <Spacer />

          <Control>
            <ControlLabel>Asset</ControlLabel>
            <ControlRow onClick={() => setIsTokenPickerOpen(true)}>
              <Input
                value={tokenInfo?.symbol || ''}
                placeholder="Token"
                icon={
                  <div>
                    {token && (
                      <Avatar
                        src={resolveUri(selectedToken?.logoURI)}
                        defaultSeed={token}
                        size={18}
                      />
                    )}
                  </div>
                }
                iconRight={<FiChevronDown size={24} />}
              />
            </ControlRow>
          </Control>
        </ControlRow>

        <TokenPicker
          // walletAddress={''}
          isOpen={isTokenPickerOpen}
          onClose={() => setIsTokenPickerOpen(false)}
          onSelect={tokenAddress => {
            console.log('tokenAddress', tokenAddress);
            setToken(tokenAddress);
            setIsTokenPickerOpen(false);
          }}
        />

        <BlockButton primary onClick={confirm} disabled={!amount || !token}>
          Approve
        </BlockButton>
      </SectionWrapper>
    </Wrapper>
  );
};

export default TokensToSpend;
