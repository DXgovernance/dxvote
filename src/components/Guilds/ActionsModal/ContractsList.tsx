import React from 'react';
import { ReactComponent as Vector } from '../../../assets/images/vector.svg';
import StyledIcon from '../common/SVG';
import {
  RegistryContract,
  useContractRegistry,
} from 'hooks/Guilds/contracts/useContractRegistry';
import { useWeb3React } from '@web3-react/core';
import { SupportedAction } from '../ActionsBuilder/types';
import {
  ActionsButton,
  ButtonDetail,
  ButtonLabel,
  SectionTitle,
  SectionWrapper,
  Wrapper,
} from './styles';

interface ContractsListProps {
  onSelect: (contract: RegistryContract) => void;
  onSupportedActionSelect: (actionType: SupportedAction) => void;
}

const ContractsList: React.FC<ContractsListProps> = ({
  onSelect,
  onSupportedActionSelect,
}) => {
  const { chainId } = useWeb3React();
  const { contracts } = useContractRegistry(chainId);

  return (
    <Wrapper>
      <SectionWrapper>
        <SectionTitle>Core</SectionTitle>
        <ActionsButton
          onClick={() =>
            onSupportedActionSelect(SupportedAction.ERC20_TRANSFER)
          }
        >
          <ButtonLabel>
            <StyledIcon src={Vector} />
            Transfer & Mint
          </ButtonLabel>
        </ActionsButton>
      </SectionWrapper>
      <SectionWrapper>
        <SectionTitle>External Contracts</SectionTitle>
        {contracts?.map(contract => (
          <ActionsButton onClick={() => onSelect(contract)}>
            <ButtonLabel>{contract.title}</ButtonLabel>
            <ButtonDetail>
              {contract.functions?.length}{' '}
              {contract.functions?.length > 1 ? 'Actions' : 'Action'}
            </ButtonDetail>
          </ActionsButton>
        ))}
      </SectionWrapper>
    </Wrapper>
  );
};

export default ContractsList;
