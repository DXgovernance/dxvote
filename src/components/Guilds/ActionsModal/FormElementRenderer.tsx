import { RegistryContractFunctionParam } from 'hooks/Guilds/contracts/useContractRegistry';
import { useMemo } from 'react';
import AddressInput from '../common/Form/AddressInput';
import { FormElementProps } from '../common/Form/common';
import DateInput from '../common/Form/DateInput';
import Input from '../common/Form/Input';
import NumericalInput from '../common/Form/NumericalInput';
import Toggle from '../common/Form/Toggle';

interface FormElementRendererProps extends FormElementProps<any> {
  param: RegistryContractFunctionParam;
}

const FormElementRenderer: React.FC<FormElementRendererProps> = ({
  param,
  value,
  onChange,
}) => {
  const FormElement: React.FC<FormElementProps<any>> = useMemo(() => {
    switch (param.component) {
      case 'address':
        return AddressInput;
      case 'integer':
      case 'decimal':
        return NumericalInput;
      case 'date':
      case 'time':
        return DateInput;
      case 'boolean':
        return Toggle;
      case 'tokenAmount':
        return NumericalInput;
      case 'contentHash':
        return Input;
      default:
        return Input;
    }
  }, [param]);

  return <FormElement value={value} onChange={onChange} />;
};

export default FormElementRenderer;
