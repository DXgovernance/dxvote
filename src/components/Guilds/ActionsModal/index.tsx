import { BigNumber, utils } from 'ethers';
import { RegistryContract } from 'hooks/Guilds/contracts/useContractRegistry';
import React, { useState } from 'react';
import { useParams } from 'react-router-dom';
import styled from 'styled-components';
import {
  defaultValues,
  getEditor,
  supportedActions,
} from '../ActionsBuilder/SupportedActions';
import {
  DecodedAction,
  DecodedCall,
  SupportedAction,
} from '../ActionsBuilder/types';
import { Button } from '../common/Button';
import { Modal } from '../common/Modal';
import ContractActionsList from './ContractActionsList';
import ContractsList from './ContractsList';
import ParamsForm from './ParamsForm';
import { useWeb3React } from '@web3-react/core';

export const EditorWrapper = styled.div`
  margin: 1.25rem;
`;

export const BlockButton = styled(Button)`
  margin-top: 1rem;
  width: 100%;
`;

interface ActionModalProps {
  isOpen: boolean;
  setIsOpen: (isOpen: boolean) => void;
  onAddAction: (action: DecodedAction) => void;
}

const ActionModal: React.FC<ActionModalProps> = ({
  isOpen,
  setIsOpen,
  onAddAction,
}) => {
  const { guild_id: guildId } = useParams<{ guild_id?: string }>();
  const { account: walletAddress } = useWeb3React();
  // Supported Actions
  const [selectedAction, setSelectedAction] = useState<SupportedAction>(null);
  const [selectedActionContract, setSelectedActionContract] =
    useState<utils.Interface>(null);

  // Generic calls
  const [selectedContract, setSelectedContract] =
    useState<RegistryContract>(null);
  const [selectedFunction, setSelectedFunction] = useState<string>(null);

  const [data, setData] = useState<DecodedCall>(null);

  function getHeader() {
    if (selectedFunction) {
      return selectedContract.functions.find(
        fn => fn.functionName === selectedFunction
      )?.title;
    }

    if (selectedContract) {
      return selectedContract?.title;
    }

    if (selectedAction) {
      return supportedActions[selectedAction].title;
    }

    return 'Add action';
  }

  function getContent() {
    if (selectedFunction) {
      const contractInterface = selectedContract.contractInterface;
      const contractId = selectedContract.contractAddress;
      return (
        <ParamsForm
          fn={selectedContract.functions.find(
            fn => fn.functionName === selectedFunction
          )}
          onSubmit={args => {
            onAddAction({
              id: `action-${Math.random()}`,
              contract: contractInterface,
              decodedCall: {
                callType: SupportedAction.GENERIC_CALL,
                from: guildId,
                to: contractId,
                function: contractInterface.getFunction(selectedFunction),
                value: BigNumber.from(0),
                args,
              },
            });
            setIsOpen(false);
          }}
        />
      );
    }

    if (selectedContract) {
      return (
        <ContractActionsList
          contract={selectedContract}
          onSelect={setSelectedFunction}
        />
      );
    }

    if (selectedAction) {
      const Editor = getEditor(selectedAction);
      return (
        <EditorWrapper>
          <Editor decodedCall={data} updateCall={setData} />
          <BlockButton onClick={saveSupportedAction}>Save Action</BlockButton>
        </EditorWrapper>
      );
    }

    return (
      <ContractsList
        onSelect={setSelectedContract}
        onSupportedActionSelect={setSupportedAction}
      />
    );
  }

  function goBack() {
    if (selectedFunction) {
      setSelectedFunction(null);
    } else if (selectedContract) {
      setSelectedContract(null);
    } else if (selectedAction) {
      setSelectedAction(null);
      setSelectedActionContract(null);
    }

    setData(null);
  }

  function setSupportedAction(action: SupportedAction) {
    const defaultDecodedAction = defaultValues[action] as DecodedAction;
    if (!defaultDecodedAction) return null;

    defaultDecodedAction.decodedCall.from = guildId;
    defaultDecodedAction.decodedCall.callType = action;
    switch (action) {
      case SupportedAction.REP_MINT:
        defaultDecodedAction.decodedCall.args.to = walletAddress;
        break;
    }
    setData(defaultDecodedAction.decodedCall);
    setSelectedAction(action);
    setSelectedActionContract(defaultDecodedAction.contract);
  }

  function saveSupportedAction() {
    if (!selectedAction || !data || !setSelectedActionContract) return;

    const decodedAction: DecodedAction = {
      id: `action-${Math.random()}`,
      decodedCall: data,
      contract: selectedActionContract,
    };

    onAddAction(decodedAction);
    setIsOpen(false);
  }

  return (
    <Modal
      isOpen={isOpen}
      onDismiss={() => setIsOpen(false)}
      header={getHeader()}
      maxWidth={300}
      backnCross={!!selectedAction || !!selectedContract}
      prevContent={goBack}
    >
      {getContent()}
    </Modal>
  );
};

export default ActionModal;
