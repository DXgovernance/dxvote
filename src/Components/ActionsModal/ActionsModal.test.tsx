import React from 'react';
import ActionsModal from './ActionsModal';
import { render } from 'utils/tests';

jest.mock('contexts/index', () => jest.fn());
jest.mock('hooks/Guilds/guild/useGuildImplementationType', () => ({
  __esModule: true,
  default: () => ({
    type: 'SnapshotERC20Guild',
    bytecode_hash:
      '0x8f98f6ff8bd58d9c9d7750c4f78c9400cf6eecbf630b94f5e481b3c7ec10ccac',
    features: ['SNAPSHOT'],
    isRepGuild: true,
    isSnapshotGuild: false,
    isSnapshotRepGuild: false,
  }),
}));

describe('ActionsModal', () => {
  let props;
  beforeEach(() => {
    props = {
      isOpen: false,
      setIsOpen: jest.fn(),
      onAddAction: jest.fn(), // (action: DecodedAction) => void;
    };
  });

  it('Should match snapshot 1', () => {
    console.error = jest.fn();
    const { container } = render(<ActionsModal {...props} />, {
      container: document.body,
    });
    expect(container).toMatchSnapshot();
  });

  it('Should match snapshot 2', () => {
    console.error = jest.fn();
    props.isOpen = true;
    jest
      .spyOn(React, 'useState')
      .mockReturnValueOnce([
        'GENERIC_CALL',
        () => console.log('setSelectedAction'),
      ]);
    const { container } = render(<ActionsModal {...props} />, {
      container: document.body,
    });
    expect(container).toMatchSnapshot();
  });
});
