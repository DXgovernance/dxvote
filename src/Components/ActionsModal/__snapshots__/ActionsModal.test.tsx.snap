// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionsModal Should match snapshot 1 1`] = `
<body>
   
   
</body>
`;

exports[`ActionsModal Should match snapshot 2 1`] = `
<body>
  <div>
    <div
      class="sc-crrszt jBqxuR"
    />
    <div
      class="sc-iqHYmW crryic"
    >
      <div
        class="sc-dQoVA jYXFLB"
      >
        
         
        <div
          class="sc-bqyKOL ATKti"
        >
          <div
            class="sc-jrAFXE sc-hHfuMS fguxYT ijBGJk"
          >
            <svg
              class="sc-dmlqKv lcxXPI"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M217.9 256L345 129c9.4-9.4 9.4-24.6 0-33.9-9.4-9.4-24.6-9.3-34 0L167 239c-9.1 9.1-9.3 23.7-.7 33.1L310.9 417c4.7 4.7 10.9 7 17 7s12.3-2.3 17-7c9.4-9.4 9.4-24.6 0-33.9L217.9 256z"
              />
            </svg>
            <h2
              class="sc-iBPTik sc-hBEYId cBZYjx eyJyRi"
            >
              Generic Call
            </h2>
          </div>
          <svg
            class="sc-fodVek jXmmJt"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </div>
        <div
          class="sc-bkzYnD jrpQyy"
        >
          <div
            class="sc-gTgzoy kFxbiQ"
            data-testid="actions-modal-editor"
          >
            <div>
              Generic Call Editor
            </div>
            <button
              class="sc-bdfBQB sc-laRQdt cnzpap erdehE"
            >
              guilds.saveAction
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
   
   
</body>
`;
